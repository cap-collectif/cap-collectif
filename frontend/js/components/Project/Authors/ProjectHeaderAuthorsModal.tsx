import * as React from 'react'
import styled from 'styled-components'
import { graphql, useFragment } from 'react-relay'
import { useIntl } from 'react-intl'
import { ListGroupItem } from 'react-bootstrap'
import { Modal, Button, Heading, CapUIModalSize } from '@cap-collectif/ui'
import ListGroupFlush from '../../Ui/List/ListGroupFlush'
import UserAvatar from '../../User/UserAvatar'
import useFeatureFlag from '@shared/hooks/useFeatureFlag'
import type { ProjectHeaderAuthorsModal_project$key } from '~relay/ProjectHeaderAuthorsModal_project.graphql'
import ResetCss from '~/utils/ResetCss'
type Props = {
  readonly show: boolean
  readonly onClose: () => void
  readonly project: ProjectHeaderAuthorsModal_project$key
}
const ProjectAuthorItem = styled(ListGroupItem).attrs({
  className: 'projectAuthorItem',
})`
  border-right: none;
  border-left: none;
`
const ProjectAuthorList = styled(ListGroupFlush)`
  .projectAuthorItem:first-of-type {
    border-top: none;
  }

  .projectAuthorItem:last-of-type {
    border-bottom: none;
  }
`
const FRAGMENT = graphql`
  fragment ProjectHeaderAuthorsModal_project on Project {
    authors {
      __typename
      ... on User {
        ...UserAvatar_user
        userType {
          name
        }
      }
      id
      url
      username
    }
  }
`

const ProjectHeaderAuthorsModal = ({ show, onClose, project }: Props): JSX.Element => {
  const profilesToggle = useFeatureFlag('profiles')
  const intl = useIntl()
  const data = useFragment(FRAGMENT, project)
  return (
    <Modal
      id="show-authors-modal"
      baseId="show-authors-modal"
      show={show}
      onClose={onClose}
      ariaLabel={intl.formatMessage({
        id: 'project.authors',
      })}
      size={CapUIModalSize.Lg}
    >
      <ResetCss>
        <Modal.Header>
          <Heading>
            {intl.formatMessage(
              {
                id: 'number-of-authors',
              },
              {
                num: data.authors ? data.authors.length : 0,
              },
            )}
          </Heading>
        </Modal.Header>
      </ResetCss>
      <Modal.Body>
        <ProjectAuthorList className="mb-0">
          {data.authors.map(user => (
            <ProjectAuthorItem key={user.id}>
              <div className="d-flex">
                <UserAvatar user={user} />
                <div className="d-flex fd-column">
                  {profilesToggle ? (
                    <a href={user.url}>{user.username}</a>
                  ) : (
                    <span className="font-weight-bold">{user.username}</span>
                  )}
                  <span className="excerpt">{user.userType ? user.userType.name : ''}</span>
                </div>
              </div>
            </ProjectAuthorItem>
          ))}
        </ProjectAuthorList>
      </Modal.Body>
      <Modal.Footer spacing={2}>
        <Button variant="primary" variantSize="medium" onClick={onClose}>
          {intl.formatMessage({
            id: 'global.close',
          })}
        </Button>
      </Modal.Footer>
    </Modal>
  )
}

export default ProjectHeaderAuthorsModal
