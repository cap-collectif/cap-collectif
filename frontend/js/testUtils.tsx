import * as React from 'react'
import ReactDOM from 'react-dom'
import { ThemeProvider } from 'styled-components'
import { Provider } from 'react-redux'
import { Provider as ReakitProvider } from 'reakit'
import { IntlProvider } from 'react-intl-redux'
import { RelayEnvironmentProvider } from 'react-relay'
import { CapUIProvider } from '@cap-collectif/ui'
import { BrowserRouter as Router } from 'react-router-dom'
import { theme } from '~/styles/theme'
import appStore from '~/stores/AppStore'
import type { FeatureFlagType } from '~relay/useFeatureFlagQuery.graphql'
import { FormProvider, useForm } from 'react-hook-form'

export const mockRandomValues = () => {
  global.Math.random = () => 0.5
}
export const enableFeatureFlags = (flags: Array<FeatureFlagType>) => {
  global.mockFeatureFlag.mockImplementation((flag: Array<FeatureFlagType>) => {
    if (flags.includes(flag)) {
      return true
    }

    return false
  })
}
export const disableFeatureFlags = () => {
  global.mockFeatureFlag.mockImplementation(() => false)
}
export const mockUrl = (url: string) => {
  // https://stackoverflow.com/questions/54021037/how-to-mock-window-location-href-with-jest-vuejs
  delete window.location
  // @ts-expect-error assign new URL(...) to window.location because property location is not writable
  window.location = new URL(url)
}
export const addsSupportForPortals = () => {
  // See: https://github.com/facebook/react/issues/11565
  // @ts-expect-error
  ReactDOM.createPortal = jest.fn(element => {
    return element
  })
}
export const clearSupportForPortals = () => {
  // @ts-expect-error
  ReactDOM.createPortal.mockClear()
}
type Props = {
  children: JSX.Element | JSX.Element[] | string
  store?: any
  useCapUIProvider?: boolean
}
export const MockProviders = ({ children, store = {}, useCapUIProvider = false }: Props) => {
  return (
    <Provider store={appStore(store)}>
      <Router>
        <IntlProvider>
          <ReakitProvider>
            {useCapUIProvider ? (
              <CapUIProvider>{children}</CapUIProvider>
            ) : (
              <ThemeProvider theme={theme}>{children}</ThemeProvider>
            )}
          </ReakitProvider>
        </IntlProvider>
      </Router>
    </Provider>
  )
}
type RelaySuspensFragmentTestProps = Props & {
  environment: any
}
export const RelaySuspensFragmentTest = ({
  children,
  environment,
  store = {},
  useCapUIProvider = false,
}: RelaySuspensFragmentTestProps) => {
  return (
    <MockProviders store={store} useCapUIProvider={useCapUIProvider}>
      <RelayEnvironmentProvider environment={environment}>
        <React.Suspense fallback="">{children}</React.Suspense>
      </RelayEnvironmentProvider>
    </MockProviders>
  )
}

export const FormWrapper = props => {
  const formMethods = useForm()

  return <FormProvider {...formMethods}>{props.children}</FormProvider>
}

export default MockProviders
