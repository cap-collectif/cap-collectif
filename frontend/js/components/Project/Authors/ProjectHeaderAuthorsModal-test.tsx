/* eslint-env jest */
import * as React from 'react'
import { graphql, useLazyLoadQuery } from 'react-relay'
import { createMockEnvironment, MockPayloadGenerator } from 'relay-test-utils'
import ReactTestRenderer from 'react-test-renderer'
import ProjectHeaderAuthorsModal from './ProjectHeaderAuthorsModal'
import { addsSupportForPortals, clearSupportForPortals, RelaySuspensFragmentTest } from '~/testUtils'
import type { ProjectHeaderAuthorsModalTestQuery } from '~relay/ProjectHeaderAuthorsModalTestQuery.graphql'

describe('<ProjectHeaderAuthorsModal />', () => {
  let environment
  let TestComponent
  let handleClose
  const query = graphql`
    query ProjectHeaderAuthorsModalTestQuery($id: ID = "<default>") @relay_test_operation {
      project: node(id: $id) {
        ...ProjectHeaderAuthorsModal_project
      }
    }
  `
  afterEach(() => {
    clearSupportForPortals()
  })
  beforeEach(() => {
    addsSupportForPortals()
    environment = createMockEnvironment()
    handleClose = jest.fn()

    const TestRenderer = props => {
      const data = useLazyLoadQuery<ProjectHeaderAuthorsModalTestQuery>(query, {})
      if (!data.project) return null
      return <ProjectHeaderAuthorsModal project={data.project} onClose={handleClose} show {...props} />
    }

    TestComponent = props => (
      <RelaySuspensFragmentTest environment={environment} useCapUIProvider>
        <TestRenderer {...props} />
      </RelaySuspensFragmentTest>
    )

    environment.mock.queueOperationResolver(operation =>
      MockPayloadGenerator.generate(operation, {
        User: () => ({
          id: 'VXNlcjp1c2VyQWRtaW4=',
          username: 'admin',
          avatarUrl: null,
          url: 'https://capco.dev/profile/admin',
          userType: {
            name: 'Citoyen',
          },
          media: null,
        }),
      }),
    )
  })
  it('renders correctly', () => {
    const wrapper = ReactTestRenderer.create(<TestComponent />)
    expect(wrapper).toMatchSnapshot()
  })
})
