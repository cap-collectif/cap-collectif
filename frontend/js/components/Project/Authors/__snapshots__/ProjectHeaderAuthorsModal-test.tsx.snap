// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectHeaderAuthorsModal /> renders correctly 1`] = `
Array [
  <div
    className="sc-bcXGCL cgAMhb toasts-container toasts-container--top"
  />,
  <div
    ariaLabel="project.authors"
    baseId="show-authors-modal"
    data-testid="Modal"
    id="show-authors-modal"
    onClose={[MockFunction]}
    show={true}
    size="lg"
  >
    <div
      className="ResetCss-jmWVRh hylUuV"
    >
      <div
        data-testid="Header"
      >
        <h2
          className="sc-bcXGCL jyqEHl cap-heading"
        >
          number-of-authors
        </h2>
      </div>
    </div>
    <div
      data-testid="Body"
    >
      <div
        className="ListGroupFlush__Container-eGHZgt cEkgdF list-group-flush ProjectHeaderAuthorsModal__ProjectAuthorList-exQxKa gZQIvY mb-0 list-group"
        striped={false}
      >
        <span
          className="ProjectHeaderAuthorsModal__ProjectAuthorItem-fAbEbJ cawNyi projectAuthorItem list-group-item"
        >
          <div
            className="d-flex"
          >
            <span
              className="UserAvatar__UserAvatarContainer-jxoNlw imfDjz"
              hasBadge={false}
            >
              <div
                className="sc-bcXGCL iLOWAE sc-ckEbfr gIIayp cap-avatar"
                title="<mock-value-for-field-\\"username\\">"
              >
                &lt;
              </div>
            </span>
            <div
              className="d-flex fd-column"
            >
              <span
                className="font-weight-bold"
              >
                &lt;mock-value-for-field-"username"&gt;
              </span>
              <span
                className="excerpt"
              >
                Citoyen
              </span>
            </div>
          </div>
        </span>
      </div>
    </div>
    <div
      data-testid="Footer"
      spacing={2}
    >
      <button
        className="sc-bcXGCL sc-laZSpB hzCEQq ktRJXy cap-button"
        onClick={[MockFunction]}
        type="button"
      >
        global.close
      </button>
    </div>
  </div>,
]
`;
